api/client/gro_client_test.py

```diff
@@ -1,92 +1,373 @@
+        self.assertTrue(isinstance(region, dict))
+        self.assertTrue('id' in region)
+        self.assertEqual(region['name'], 'United States')
+
+    def test_lookup_belongs(self):
+        self.assertEqual(next(self.client.lookup_belongs('regions', 1215))['name'], 'World')
+
+    def test_rank_series_by_source(self):
+        series = next(self.client.rank_series_by_source([]))
+        self.assertTrue('metric_id' in series)
+        self.assertTrue('source_id' in series)
+
+    def test_get_geo_centre(self):
+        centre = self.client.get_geo_centre(1215)
+        self.assertTrue(len(centre) == 1)
+        self.assertTrue('centre' in centre[0])
+        self.assertTrue('regionName' in centre[0])
+
+    def test_get_geojson(self):
+        geojson = self.client.get_geojson(1215)
+        self.assertTrue('type' in geojson)
+        self.assertTrue('type' in geojson['geometries'][0])
+        self.assertTrue('coordinates' in geojson['geometries'][0])
+
+    def test_get_descendant_regions(self):
+        self.assertTrue('name' in self.client.get_descendant_regions(1215)[0])
+        self.assertTrue('name' not in self.client.get_descendant_regions(1215, include_details=False)[0])
+
+    def test_get_available_timefrequency(self):
+        self.assertEqual(self.client.get_available_timefrequency()[0]['frequency_id'], 3)
+
+    def test_get_top(self):
+        self.assertEqual(self.client.get_top('regions',
+                                             metric_id=860032,
+                                             item_id=274,
+                                             frequency_id=9,
+                                             source_id=2)[0]['regionId'], 1215)
+
+    def test_get_df(self):
+        client = GroClient(MOCK_HOST, MOCK_TOKEN)
+        client.add_single_data_series(mock_data_series[0])
+        df = client.get_df()
+        self.assertEqual(df.iloc[0]['start_date'].date(), date(2017, 1, 1))
+        client.add_single_data_series(mock_data_series[0])
+        df = client.get_df(show_revisions=True)
+        self.assertEqual(df.iloc[0]['start_date'].date(), date(2017, 1, 1))
+        indexed_df = client.get_df(index_by_series=True)
+        self.assertEqual(indexed_df.iloc[0]['start_date'].date(), date(2017, 1, 1))
+        series = dict(zip(indexed_df.index.names, indexed_df.iloc[0].name))
+        self.assertEqual(series, mock_data_series[0])
+
+    def test_get_df_show_revisions(self):
+        client = GroClient(MOCK_HOST, MOCK_TOKEN)
+        client.add_single_data_series(mock_data_series[0])
+        df = client.get_df(show_revisions=True)
+        self.assertEqual(df.iloc[0]['start_date'].date(), date(2017, 1, 1))
+
+    def test_add_points_to_df(self):
+        client = GroClient(MOCK_HOST, MOCK_TOKEN)
+        client.add_points_to_df(None, mock_data_series[0],
+                                client.get_data_points(**mock_data_series[0]))
+        df = client.get_df()
```

Comment:
minor. use a mock data frame? this fn is also tested in this module

Reply:
in general some of these nested tested client fn should be mocked when being called by another. ie gdh returns client.get_df, are we not doing any stubs or mocking client.get_df for this case. i guess it works for now since these are also tested in this module, but if they require changes, we would need to update two unite tests

Reply:
Yeah, fair. This is blending unit/integration testing. Will split up

Reply:
Okay so I've been thinking about this a bit: for `GDH()`, I separated out the parsing logic into its own util now and tested that separately from the `GDH()` function itself. So that's a bit more of a "unit" now that is unit tested.

In terms of mocking `get_df()` in `GDH()` though, I think the tests should know as little as possible about how the methods are implemented - it shouldn't matter if it calls `get_df()` or not, from the test's perspective. I just want a certain output for my given input. What if the method's changed to use something else internally but the output's the same? That's a fine change - maintains backwards compatibility and everything but is just an internal refactor - but my test would then be invalid.

Similarly, with `add_points_to_df()`, I think the expected behavior is that I add the points using this method and then can get the points back by using `get_df()`. If anything about that usage changes, that seems like a compatibility issue. Not sure I want to be inspecting the internal `_data_frame` property, since that's not one of the public interfaces that's guaranteed not to change - `get_df()` is.

Reply:
fair, like i said before, we also tested the get_df fn so i think its reasonable to use here. i think theres arguments to both. now, if get_df is updated this might change the validity for both unit tests. im not sure why you mentioned that get_df is guaranteed to not change.